libdevice/device.h

```diff
@@ -9,6 +9,10 @@
 #ifndef __LIBDEVICE_DEVICE_H__
 #define __LIBDEVICE_DEVICE_H__
 
+// We need the following header provided by the sycl project to ensure
+// definition of all spirv variables required by the wrapper libraries.
+#include "spirv_vars.hpp"
```

Comment:
This looks like a broken design.
`libdevice` should not include `sycl` library headers.
@vzakhari, recently moved this library out of the `sycl` project to make them independent (i.e. we should be able to one w/o the other).

Doesn't #1384 help to have all needed definitions?

Reply:
I see that #1384 removed all __spirv_* declarations from libdevice/device_math.h, and I do not understand how it passed the testing.  @bader, do you know?

I agree that devicelib must not depend on any headers from SYCL project.  We are building devicelib in a fork without SYCL support, so I'd rather keep devicelib self-contained.

Reply:
> I see that #1384 removed all __spirv_* declarations from libdevice/device_math.h, and I do not understand how it passed the testing. @bader, do you know?
> 
> I agree that devicelib must not depend on any headers from SYCL project. We are building devicelib in a fork without SYCL support, so I'd rather keep devicelib self-contained.

With [this change](https://github.com/intel/llvm/pull/1374/) SPIR-V built-ins are "clang" built-ins i.e. they are recognized by clang w/o forward declarations as they are declared by clang itself.

Reply:
Thank you!  I will have to use -fdeclare-spirv-builtins for non-SYCL builds of libdevice.

Reply:
> 
> 
> This looks like a broken design.
> `libdevice` should not include `sycl` library headers.
> @vzakhari, recently moved this library out of the `sycl` project to make them independent (i.e. we should be able to one w/o the other).
> 
> Doesn't #1384 help to have all needed definitions?

This is required because the `_devicelib_assert_fail` calls in glibc_wrapper.cpp and msvc_wrapper.cpp both require the global and local invocation IDs.
```
  __devicelib_assert_fail(
      expr, file, line, func, __spirv_GlobalInvocationId_x(),
      __spirv_GlobalInvocationId_y(), __spirv_GlobalInvocationId_z(),
      __spirv_LocalInvocationId_x(), __spirv_LocalInvocationId_y(),
      __spirv_LocalInvocationId_z());
```

The upstream llvm-spirv translator doesn't implement these, so we implemented them in a somewhat broken way in the llvm-spirv translator, see https://github.com/intel/llvm/issues/1166. This patch is triyng to avoid upstreaming the broken changes to the translator by implementing the invocation ID functions (among some others) for non NVPTX inside spirv_vars.hpp, which is why the devicelib has a requirement on the SYCL header.

If you'd prefer, an alternative to the header dependency would be to have another header in the devicelib which implements only the required invocation ID functions.

Reply:
Yes, please make the implementation in a devicelib header.  Not that I like it, but I want devicelib to be self-contained, and do not see a better way now.

Reply:
Done

Reply:
Thank you, @Alexander-Johnston!  LGTM for libdevice.